src/Jinja2CppLight.h

```diff
@@ -153,7 +153,7 @@ class Template {
 
     ValueMap valueByName;
 //    std::vector< std::string > varNameStack;
-    Root *root;
+    std::unique_ptr<Root> root;
```

Comment:
On concern about std::unique_ptr here. It makes Template class non-copyable (due to traits of std::unique_ptr), only moveable. I'm not sure about typical usecases, but this is significant restriction. May by std::shared_ptr could be better in _this_ place.

Reply:
What if a custom copy constructor just made a new Root on copy? The root member isn't used until rendering anyway.

Reply:
In this case you need to do the deep copy rather than doing shallow copy with shared_ptr. The better solutions could be the following: remove 'root' from the template and leave it inside 'render' method. Together with 'valueByName'. The goal: to make interface completely stateless.  Or do it another way: separate rendering and parsing procedures. Anyway, I see no reason to parse template each time you render it.

Reply:
Well if we just make a new (empty) root on copy *instead* of deep copying, than the copied version will regenerate the root on render. Which practically makes no difference, since as far as I can tell `render()` re-parses every time regardless. That's probably an issue for a separate PR.

I don't think `shared_ptr` is appropriate, since we probably don't want the root shared between different template instances. Doing so would create a race condition on the data if someone runs the copied Template concurrently with the original.

Reply:
> Well if we just make a new (empty) root on copy instead of deep copying, than the copied version will regenerate the root on render. 

But in this case it wont a copy... :) 

> I don't think shared_ptr is appropriate, since we probably don't want the root shared between different template instances.

Yes. you absolutely right.

> Doing so would create a race condition on the data if someone runs the copied Template concurrently with the original.

As far, as I see, the current implementation is absolutely singlethreaded. Try to use the same template from the different threads and you will get a lot of race conditions. Both on access to the root and access to the valueByName.
So, you right. Implementing copy/move constructor as you suggest is a proper way for this implementation.